File: train/verify_demo.mjs
/*
 * Does staging/ncpu.html run the trained world, or something close to it?
 *
 *   node train/verify_demo.mjs > train/demo_run.json
 *   python3 train/verify_demo.py
 *
 * Steps the page in a real browser on a fixed sample and dumps the pointer
 * channel and the region vote, so the PyTorch port can be held against it. The
 * page reimplements the law in plain JavaScript -- the static channels are
 * convolved once and the crowding blur is folded into a single convolution --
 * and both of those are only sound if the numbers come out the same.
 */
import { chromium } from "playwright";
import { createServer } from "node:http";
import { readFile } from "node:fs/promises";
import { extname, join, resolve } from "node:path";

const ROOT = resolve(import.meta.dirname, "..");
const CHROME = "/opt/pw-browsers/chromium-1194/chrome-linux/chrome";
const STEPS = (process.env.STEPS ?? "1,4,16,64").split(",").map(Number);
const SAMPLE = Number(process.env.SAMPLE ?? 0);

const MIME = { ".html":"text/html", ".js":"text/javascript", ".json":"application/json" };
const server = createServer(async (req, res) => {
  const p = join(ROOT, decodeURIComponent(req.url.split("?")[0]).replace(/^\/+/, ""));
  // read BEFORE writing headers: a miss used to throw after the 200 was already
  // on the wire, and the catch then tried to send a second set
  let body;
  try { body = await readFile(p); }
  catch { res.writeHead(404); res.end("no"); return; }
  res.writeHead(200, { "content-type": MIME[extname(p)] ?? "application/octet-stream" });
  res.end(body);
});
await new Promise(r => server.listen(0, "127.0.0.1", r));

const browser = await chromium.launch({ executablePath: CHROME });
const page = await browser.newPage();
page.on("pageerror", e => console.error("page error:", e.message));
await page.goto(`http://127.0.0.1:${server.address().port}/staging/ncpu.html`,
                { waitUntil: "domcontentloaded" });
await page.waitForFunction(() => window.__ncpu, null, { timeout: 30000 });

const out = await page.evaluate(({ STEPS, SAMPLE }) => {
  window.__ncpu.setSample(SAMPLE);
  const res = { sample: SAMPLE, frames: [] };
  let done = 0;
  for (const s of STEPS) {
    window.__ncpu.tick(s - done); done = s;
    res.frames.push({ step: s, scores: window.__ncpu.scores(),
                      mass: window.__ncpu.mass(),
                      pointer: window.__ncpu.pointer() });
  }
  return res;
}, { STEPS, SAMPLE });

console.log(JSON.stringify(out));
await browser.close();
server.close();

File: train/verify_demo.py
"""
Hold the browser demo against the PyTorch port, step for step.

    node train/verify_demo.mjs > train/demo_run.json
    python3 train/verify_demo.py

The page takes two shortcuts the trainer does not: the static channels are
convolved once instead of every tick, and the crowding term is one blur of the
summed density instead of one per channel. Both are meant to be exact, and this
is what says whether they are. It also checks mass conservation, which is the
property the whole design rests on.
"""
import json, os, sys
import numpy as np, torch

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import digits as dg, fieldlife as fl
from train_digits import Task, roll
from eval_digits import Replay

HERE = os.path.dirname(os.path.abspath(__file__))
run = json.load(open(os.path.join(HERE, "demo_run.json")))
data = open(os.path.join(HERE, "..", "staging", "ncpu-data.js")).read()
D = json.loads(data[data.index("{"):data.rindex(";")])

cfg = json.load(open(os.path.join(HERE, "runs", D["name"], "preset-best.json")))
C = cfg["C"]
task = Task(C, D["nClass"], D["grid"], D["digitPx"], D["nStatic"], D["ring"])
world = Replay(cfg, C)

siren = None
ck = os.path.join(HERE, "runs", D["name"], "ckpt.pt")
st = torch.load(ck, weights_only=False)
if st.get("siren"):
    siren = dg.SirenSeed(D["grid"], D["nChem"], 32, 2)
    siren.load_state_dict(st["siren"])

x, y = dg.load("test")
i = run["sample"]
seeds, _, lab = task.build(x[i:i + 1], y[i:i + 1], np.random.default_rng(0), siren)

print(f"sample {i}, a {int(lab[0])}; {len(run['frames'])} checkpoints\n")
print(f"{'step':>5} {'vote max|diff|':>15} {'pointer max|diff|':>18} "
      f"{'argmax':>8} {'mass drift':>12}")
worst = 0.0
rho, done = seeds.clone(), 0
m0 = float(rho[0, task.ptr].sum())
for f in run["frames"]:
    with torch.no_grad():
        rho = roll(world, task, rho, f["step"] - done)
    done = f["step"]
    s_t = task.scores(rho)[0].numpy()
    s_j = np.array(f["scores"])
    p_t = rho[0, task.ptr].numpy().ravel()
    p_j = np.array(f["pointer"])
    dv, dp = np.abs(s_t - s_j).max(), np.abs(p_t - p_j).max()
    worst = max(worst, float(dv))
    agree = "same" if s_t.argmax() == s_j.argmax() else "DIFFER"
    drift = abs(float(np.array(f["mass"])[task.ptr]) - m0) / max(m0, 1e-9)
    print(f"{f['step']:>5} {dv:>15.3e} {dp:>18.3e} {agree:>8} {drift:>12.2e}")
print(f"\nworst vote difference {worst:.3e}  ->  "
      + ("the page runs the trained world" if worst < 1e-4
         else "THE PAGE IS RUNNING SOMETHING ELSE"))
